app/src/backlight.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+
+#include <drivers/led.h>
+#include <logging/log.h>
+#include <settings/settings.h>
+
+#include <zmk/backlight.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/activity_state_changed.h>
+#include <zmk/events/usb_conn_state_changed.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define BACKLIGHT_LABEL DT_LABEL(DT_CHOSEN(zmk_backlight))
+#define BACKLIGHT_LEVELS CONFIG_ZMK_BACKLIGHT_LEVELS
+
+struct backlight_state {
+    uint8_t brightness;
+    bool on;
+};
+
+static const struct device *led_bl;
+
+static struct backlight_state state;
+
+static int zmk_backlight_update_on() { return state.on ? led_on(led_bl, 0) : led_off(led_bl, 0); }
+
+static int zmk_backlight_update_brt() {
+    uint8_t percent = state.brightness * 100 / (BACKLIGHT_LEVELS - 1);
+    return led_set_brightness(led_bl, 0, percent);
+}
+
+static int zmk_backlight_update() {
+    int rc = zmk_backlight_update_on();
+    if (rc != 0) {
+        return rc;
+    }
+
+    rc = zmk_backlight_update_brt();
+    if (rc != 0) {
+        return rc;
+    }
+
+    return 0;
+}
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+static int bl_settings_set(const char *name, size_t len, settings_read_cb read_cb, void *cb_arg) {
+    const char *next;
+
+    if (settings_name_steq(name, "state", &next) && !next) {
+        if (len != sizeof(state)) {
+            return -EINVAL;
+        }
+
+        int rc = read_cb(cb_arg, &state, sizeof(state));
+        if (rc < 0) {
+            return rc;
+        }
+
+        return zmk_backlight_update();
+    }
+
+    return -ENOENT;
+}
+
+static struct settings_handler bl_conf = {.name = "backlight", .h_set = bl_settings_set};
+
+static void zmk_backlight_save_state_work() {
+    settings_save_one("backlight/state", &state, sizeof(state));
+}
+
+static struct k_delayed_work backlight_save_work;
+#endif // IS_ENABLED(CONFIG_SETTINGS)
+
+static int zmk_backlight_save_state() {
+#if IS_ENABLED(CONFIG_SETTINGS)
+    return settings_save_one("backlight/state", &state, sizeof(state));
```

Comment:
Instead of saving settings directly, could you debounce the save similarly to how it's done on these two lines in RGB underglow? https://github.com/zmkfirmware/zmk/blob/main/app/src/rgb_underglow.c#L261

The reason for this is to reduce flash operations. People often will be cycling backlight up, down, on, and off. This results in a pretty big amount of flash wear, and only saving after a minute of no changes will help us to reduce that.